### PR TITLE
Improve: obey module-item-spacing for let-and bindings

### PR DESCRIPTION
**Pull request by gpetiot**
Should fix issue #363

### CODE REVIEW

**Comment by hhugo** [comment]
Can you add some tests with doc comments between let-and constructs
```
let x = x

(** comment *)
and y = y
``` 

```
let x = x

(** floating comment *)

and y = y
```

**Comment by hhugo** [comment]
Also some test with multi-lines
```
let x = x
and y = something
   long
```

```
let y = something
   long
and x = x
```

**Comment by gpetiot** [comment]
I added more tests, there are some `and` that are replaced by `let`, I don't know if I messed up with the first/last variables.

File: src/Ast.ml [hunk 1]
```diff
@@ -335,6 +335,113 @@ module Signature_item : Module_item with type t = signature_item = struct
...
+    | Ppat_extension _ -> false
+    | Ppat_open (_, p) -> pattern_has_doc p
+
+  let is_simple (i, c) =
+    Poly.(c.Conf.module_item_spacing = `Compact)
+    && Location.width i.pexp_loc <= c.Conf.margin
+    && Location.is_single_line i.pexp_loc
+
+  let break_between (i1, c1) (i2, c2) =
+    expr_has_doc i1 || expr_has_doc i2
```
Comment (jberdine) [inline src/Ast.ml]: Do you have examples where checking for docs here is necessary? It is not obvious to me that checking `pvb_attributes` is not enough.

**Comment by gpetiot** [comment]
I added some commits to take attributes into account, it looks good but it required to print warnings only once to not have extra regressions

**Comment by gpetiot** [comment]
Warnings are printed when we iterate on exp attributes for update_config.

**Comment by hhugo** [comment]
Actually, ignore what I said. I wonder if the proper fix should just be to not call update_config multiple time though.

**Comment by jberdine** [comment]
@hhugo, to avoid calling update_config multiple times, do you have in mind to modify fmt_value_binding to assume the config it receives has already been updated with attributes, or something else?

File: src/Fmt_ast.ml [hunk 2]
```diff
@@ -3682,16 +3682,33 @@ and fmt_structure_item c ~last:last_item ?ext {ctx; ast= si} =
...
+      let grps =
+        List.group bindings ~break:(fun (itmI, cI) (itmJ, cJ) ->
+            (not (List.is_empty itmI.pvb_attributes))
+            || (not (List.is_empty itmJ.pvb_attributes))
+            || Ast.break_between
+                 (Exp itmI.pvb_expr, cI.conf)
+                 (Exp itmJ.pvb_expr, cJ.conf) )
+      in
+      let fmt_grp ~first:first_grp ~last:_ bindings =
+        list_fl bindings (fun ~first ~last:_ (binding, _) ->
```
Comment (hhugo) [inline src/Fmt_ast.ml]: should we make use of the config here ?

**Comment by hhugo** [comment]
@jberdine, either that or an optional argument in `update_config` to mutes the warning.

**Comment by jberdine** [comment]
This fails:
```
$ cat test.ml
(** doc *)

type t
$ ocamlformat --break-struct=natural test.ml
ocamlformat.exe: Cannot process "test.ml".
  Please report this bug at https://github.com/ocaml-ppx/ocamlformat/issues.
  BUG: formatting changed ast.
[Exit 1]
```

**Comment by gpetiot** [comment]
Are you sure this is due to this PR ?
I will try to fix it as well as other issues in https://github.com/ocaml-ppx/ocamlformat/pull/380

**Comment by jberdine** [comment]
Yes, this fails with f516634 but not with a1bbe9d

**Comment by gpetiot** [comment]
Exactly this commit belongs to this other PR.

**Comment by jberdine** [comment]
Ah, ok, thanks, I must be confused about which commits are where. I'll look around.

**Comment by jberdine** [comment]
Ok, sorry, I was confused, you're right that this one is ok. BTW, I'm not sure why but the title of the commit didn't end up being synced with the title of the PR.

**Pull request merged**